Enable validation for jira issue status

NasAmin opened the pull request:
Enables validation of jira issue status against a list of allowed statuses provided as an action input.
This implements #32 
Please refer to #32 for more details.

Commit:
Added status field to the JiraDetails object. Updated tests for the new field

Commit:
Added support for validating jira issue status based on allowed list. Fixed eslint issues

Commit:
Added issue status validation and also added status to the PR description

Commit:
Tidied up flag names

Commit:
Updated documentation

NasAmin commented:
Please review.
Apologies if I have missed anything and please let me know if this can be done in a better way.

Commit:
Removed duplication option

Commit:
Fxed typo in readme

Commit:
Fixed isIssueStatusInvalid condition

NasAmin commented:
I'll appreciate if the maintainers can review this PR please

rheaditi commented:
Hey @NasAmin! Thanks for your contribution & raising this PR! Really glad you found this action useful :)

We've had issues with the markdown/markup syntax not getting reproduced correctly in the GitHub PR description. Could you add some screenshots of this in a test repo with the working PR description? Unfortunately we only have this means of manual testing..

---

@loftykhanna: Could you have a 👀 on this PR, please?

NasAmin commented:
@rheaditi Thanks for getting back to me.
The PR description issue, is this a general issue or are you talking the markup this PR will be generating?
I have put an the generated markup on issue #32 
This is what my PR will generate:
![image](https://user-images.githubusercontent.com/12657747/97967482-2d3a1780-1db5-11eb-925e-eebcddd0ab0c.png)

If you are talking about the general PR description bit of jira-lint then I can look into that but it should probably be a separate PR.

Commit:
Merged from upstream and resolved merge conflicts for index.js

Commit:
Regenerated index.js after merging from upstream

NasAmin commented:
@loftykhanna @maddhruv I would really appreciate if you can review and approve this PR and release it.

rheaditi commented on src/main.ts:
```diff
@@ -129,6 +144,18 @@ async function run(): Promise<void> {
+          body: getInvalidIssueStatusComment(details.status, ALLOWED_ISSUE_STATUSES),
+        };
+        console.log('Adding comment for invalid issue status');
+        addComment(client, invalidIssueStatusComment);
```
@NasAmin you may need an await here..

loftykhanna reviewed (changes requested):
@NasAmin I tested your code, no comments are getting added after it fails.
@rheaditi Comment/suggestion might help in fixing this. Please fix and I will test again.

Also, a suggestion, can 'allowed_issue_statuses' be case insensitive?
![image](https://user-images.githubusercontent.com/10628479/101484207-423f3300-397f-11eb-8897-18ad0255ee42.png)

rheaditi commented on README.md:
```diff
@@ -96,6 +96,33 @@ When a PR passes the above check, `jira-lint` will also add the issue details to
 
+#### Issue Status Validation
+Issue status is shown in the [Description](#description).
+**Why validation issue status?** 
```
```suggestion
**Why validate issue status?** 
```

rheaditi commented on README.md:
```diff
@@ -96,6 +96,33 @@ When a PR passes the above check, `jira-lint` will also add the issue details to
+#### Issue Status Validation
+Issue status is shown in the [Description](#description).
+**Why validation issue status?** 
+In some cases, you may be pushing changes for a story that is set to `Done` or it may not have been pulled into working backlog. This section describes how we can help engineers avoid pushing to stories that are set to invalid statuses.
```
```suggestion
In some cases, you may be pushing changes for a story that is set to `Done` or it may not have been pulled into working backlog. This section describes how we can help engineers avoid pushing to branches for stories that are set to statuses other than the ones allowed in the project; for example - you may want to allow PRs for stories that are in `To Do`/`Planning`/`In Progress` states.
```

rheaditi commented on README.md:
```diff
@@ -96,6 +96,33 @@ When a PR passes the above check, `jira-lint` will also add the issue details to
+  - This will only be used when `validate_issue_status` is `true`. This should be a comma separated list of statuses. If the detected issue's status is not in one of the `allowed_issue_statuses` then `jira-lint` will fail the status check.
+
+**Example of invalid status**
+  <p>:broken_heart: The detected issue is not in one of the allowed statuses :broken_heart: </p>    
```
:+1: For the format of this comment :D

rheaditi commented on README.md:
```diff
@@ -140,11 +167,14 @@ When a PR passes the above check, `jira-lint` will also add the issue details to
 | `skip-branches` | A regex to ignore running `jira-lint` on certain branches, like production etc.                                                                                                                                                                                                                                    | false    | ' '     |
 | `skip-comments` | A `Boolean` if set to `true` then `jira-lint` will skip adding lint comments for PR title.                                                                                                                                                                                                                         | false    | false   |
 | `pr-threshold`  | An `Integer` based on which `jira-lint` will add a comment discouraging huge PRs.                                                                                                                                                                                                                                  | false    | 800     |
+| `validate_issue_status`  | An `Boolean` based on which `jira-lint` will validate the status of the detected jira issue                                                                                                                                                                                                              | false    | false   |
```
```suggestion
| `validate_issue_status`  | A `Boolean` based on which `jira-lint` will validate the status of the detected jira issue                                                                                                                                                                                                              | false    | false   |
```

rheaditi commented on README.md:
```diff
@@ -140,11 +167,14 @@ When a PR passes the above check, `jira-lint` will also add the issue details to
 | `skip-comments` | A `Boolean` if set to `true` then `jira-lint` will skip adding lint comments for PR title.                                                                                                                                                                                                                         | false    | false   |
 | `pr-threshold`  | An `Integer` based on which `jira-lint` will add a comment discouraging huge PRs.                                                                                                                                                                                                                                  | false    | 800     |
+| `validate_issue_status`  | An `Boolean` based on which `jira-lint` will validate the status of the detected jira issue                                                                                                                                                                                                              | false    | false   |
+| `allowed_issue_statuses`  | An comma separated string. The detected jira issue's status will be compared against this list and if a match is not found then the status check will fail. Requires `validate_issue_status` to be set to `true`.                                                                                        | false    | In Progress |
```
```suggestion
| `allowed_issue_statuses`  | A comma separated list of allowed statuses. The detected jira issue's status will be compared against this list and if a match is not found then the status check will fail. *Note*: Requires `validate_issue_status` to be set to `true`.                                                                                        | false    | `"In Progress"` |
```

rheaditi commented on src/utils.ts:
```diff
@@ -321,3 +337,44 @@ Valid sample branch names:
+    return true;
+  }
+
+  if (allowedIssueStatuses.includes(details.status)) {
+    return true;
+  } else {
+    return false;
+  }
```
[nit]

```suggestion
  return allowedIssueStatuses.includes(details.status);
```

rheaditi reviewed (changes requested):
@NasAmin have a few minor comments.. Please take a look whenever you get time 🙂

Commit:
Added await to async call for adding a PR comment

NasAmin committed:
Updated readme based on suggested changes on PR review

Co-authored-by: Aditi Mohanty <rheaditi@users.noreply.github.com>

NasAmin committed:
Update README.md - applied suggested markup changes from PR review

Co-authored-by: Aditi Mohanty <rheaditi@users.noreply.github.com>

NasAmin committed:
Update README.md - Fixed typo suggested during PR review

Co-authored-by: Aditi Mohanty <rheaditi@users.noreply.github.com>

NasAmin committed:
Fixed allowed_issue_statuses description from PR review

Co-authored-by: Aditi Mohanty <rheaditi@users.noreply.github.com>

NasAmin committed:
Shortened if conditions based on PR review suggestion

Co-authored-by: Aditi Mohanty <rheaditi@users.noreply.github.com>

Commit:
Removed duplicate status field from issue status markup

NasAmin commented on src/utils.ts:
```diff
@@ -263,6 +275,10 @@ export const getPRDescription = (body = '', details: JIRADetails): string => {
       <th>Labels</th>
       <td>${getLabelsForDisplay(details.labels)}</td>
     </tr>
+    <tr>
```
Good catch!
Fixed.

NasAmin commented:
> @NasAmin I tested your code, no comments are getting added after it fails.
> @rheaditi Comment/suggestion might help in fixing this. Please fix and I will test again.
> 
> Also, a suggestion, can 'allowed_issue_statuses' be case insensitive?
> ![image](https://user-images.githubusercontent.com/10628479/101484207-423f3300-397f-11eb-8897-18ad0255ee42.png)

@loftykhanna Thanks for testing this. Not sure how I can easily test the PR comment bit if it fails but it might have been because the `addComment()` call didn't have an `await`. @rheaditi suggested that change which I have applied.

Regarding `allowed_issue_statuses` case insensitive, the statuses are passed to Jira really so if Jira supports case insensitive it should work otherwise we may not have a choice. But AFAIK, JQL is case insensitive. I just tested with the following
![image](https://user-images.githubusercontent.com/12657747/101625189-fd64dc00-3a12-11eb-8149-6b7dea8f5486.png)

NasAmin commented:
@rheaditi and @loftykhanna thank you both for reviewing my code. I have addressed all the comments (I think).
However, I am not sure how to test the PR comment if the issue status check fails. @loftykhanna Could you please test it again or let me know how to test it?

If there is a way to write some kind of automated test then that would be great for the future.

loftykhanna commented:
@NasAmin Sure let me test and get back to you on this.

loftykhanna commented:
@NasAmin Thank you for your contribution, Its working fine now 👍
